### ISSUE TITLE
httpd: enable support for no HTTP/2

### ISSUE THREAD

**Issue by dilyanpalauzov** [cyrusimap/cyrus-imapd#2401]
* ./configure --without-nghttp2 disables HTTP/2 even when libnghttp2 is installed
* when HTTP/2 is not enabled/installed, don't send
>          Connection: Upgrade
>          Upgrade:
headers
* I am not sure if the changes in imap/httpd.c are correct, in particular how to handle `UPGRADE_TLS` .
```diff
diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -1448,7 +1448,6 @@ AC_ARG_ENABLE(http,
         [AS_HELP_STRING([--enable-http], [enable HTTP support])],, [enable_http="no";])
 AM_CONDITIONAL([HTTPD], [test "$enable_http" != "no"])
 
-with_nghttp2=no
 with_xml2=no
 with_ical=no
 with_icu4c=no
@@ -1558,14 +1557,20 @@ dnl                        AC_DEFINE(WITH_DKIM,[],
 dnl                                [Build DKIM support into iSchedule?]),
 dnl                        AC_MSG_WARN([Your version of OpenDKIM can not support iSchedule.  Consider patching OpenDKIM with contrib/dkim_canon_ischedule.patch])),
 dnl                AC_MSG_WARN([Your version of OpenDKIM can not support iSchedule.  Consider upgrading to OpenDKIM >= 2.7.0]))
-
-        PKG_CHECK_MODULES([NGHTTP2], [libnghttp2 >= 1.5], [
-                AC_DEFINE(HAVE_NGHTTP2,[],
-                        [Build HTTP/2 support into httpd?])
-                with_nghttp2=yes
-                ],
-                AC_MSG_NOTICE([httpd will not have support for HTTP/2.  Consider installing libnghttp2 >= 1.5]))
-
+        AC_ARG_WITH(nghttp2, [AS_HELP_STRING([--without-nghttp2], [disable HTTP/2 support (check)])],,[with_nghttp2="check"])
+        if test "x$with_nghttp2" = "xyes" -o "x$with_nghttp2" = "xcheck"; then
+                PKG_CHECK_MODULES([NGHTTP2], [libnghttp2 >= 1.5], [
+                    with_nghttp2=yes
+                    AC_DEFINE(HAVE_NGHTTP2,[],
+                            [Build HTTP/2 support into httpd?])
+                ], [
+                    if test "x$with_nghttp2" = "xyes"; then
+                        AC_MSG_ERROR([HTTP/2 explicitly requested, but libnghttp2 was not found])
+                    fi
+                    with_nghttp2=no
+                    AC_MSG_NOTICE([httpd will not have support for HTTP/2.  Consider installing libnghttp2 >= 1.5])
+                ])
+        fi
 
         PKG_CHECK_MODULES([BROTLI], [libbrotlienc], [
                 AC_DEFINE(HAVE_BROTLI,[],
diff --git a/imap/httpd.c b/imap/httpd.c
--- a/imap/httpd.c
+++ b/imap/httpd.c
@@ -1621,10 +1621,14 @@ static int examine_request(struct transaction_t *txn)
     }
     else if (!httpd_tls_done && txn->flags.ver == VER_1_1) {
         /* Advertise available upgrade protocols */
+#ifdef HAVE_NGHTTP2
         txn->flags.conn |= CONN_UPGRADE;
         txn->flags.upgrade = UPGRADE_HTTP2;
-        if (config_mupdate_server && config_getstring(IMAPOPT_PROXYSERVERS))
+#endif
+        if (config_mupdate_server && config_getstring(IMAPOPT_PROXYSERVERS)) {
             txn->flags.upgrade |= UPGRADE_TLS;
+            txn->flags.conn |= CONN_UPGRADE;
+        }
     }
 
     query = URI_QUERY(txn->req_uri);
```

**Comment by ksmurchison** [comment]
What's the point of explicitly disabling HTTP/2?  If the client doesn't support it, the Upgrade header doesn't hurt anything.  If the client does support it, its a more efficient use of the TCP connection and is a win for both client and server.

**Comment by dilyanpalauzov** [comment]
I want to run cyrus-httpd as backend and Nginx*proxy as front end.  Nginx handles both TLS and HTTP/2 with he clients.  Hence cyrus-httpd shall not send to Nginx `Connect:Upgrade, Upgrade: ""`.  The latter forwards them to the clients further and most clients just work, but curl with libnghttp2 support states, that HTTP/2 connection cannot be established, so I had to call  `curl --http1.1`.

**Comment by ksmurchison** [comment]
That's a bug in Nginx.  It should not be forwarding those headers (see https://tools.ietf.org/html/rfc7230#section-6.1)

And don't you want Nginx to have the ability to connect to Cyrus via HTTP/2?

**Comment by dilyanpalauzov** [comment]
I am not sure if Nginx forwards the headers, but without libnghttp2 and without the changes in imap/httpd.c I get:
```
$ curl -D- http://127.0.0.3/freebusy/user/me
HTTP/1.1 200 OK
Date: Wed, 20 Jun 2018 19:07:21 GMT
Connection: Upgrade
Upgrade: 
Cache-Control: no-transform
Vary: Accept-Encoding
Accept-Ranges: none
Content-Type: text/calendar; charset=utf-8
Content-Disposition: inline; filename="me.ifb"
Content-Length: 319

BEGIN:VCALENDAR
VERSION:2.0
```
-> `Upgrage:` without anything after it.

And then `$curl -D- -v https://mail.aegee.org/user/freebusy/me` prints:
```
*   Trying 144.76.142.78...
* TCP_NODELAY set
* Connected to mail.aegee.org (144.76.142.78) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=lists.aegee.org
*  start date: Apr  8 08:57:55 2018 GMT
*  expire date: Jul  7 08:57:55 2018 GMT
*  subjectAltName: host "mail.aegee.org" matched cert's "mail.aegee.org"
*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x1bede90)
> GET /freebusy/user/me HTTP/2
> Host: mail.aegee.org
> User-Agent: curl/7.60.0
> Accept: */*
> 
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
* http2 error: Invalid HTTP header field was received: frame type: 1, stream: 1, name: [upgrade], value: []
* HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
* Connection #0 to host mail.aegee.org left intact
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
```
So the question is first, if httpd can sent Upgrade: without anything and second, if nginx is allowed to forward the Upgrade headers.

I don't want to use the ability to connect Nginx with httpd over HTTP/2, as both processes are on the same host and this additional encoding/decoding to HTTP/2 just slows everything down.

**Comment by ksmurchison** [comment]
OK.  The empty Upgrade is a Cyrus bug.  I will fix that.

**Comment by ksmurchison** [comment]
Actually, I can't reproduce the empty Upgrade header with the code in the master branch.  What version are you using?

**Comment by dilyanpalauzov** [comment]
I use 3.0, compare the changes in.imap/httpd.c: instable:
```c
    else if (!txn->conn->tls_ctx && txn->flags.ver == VER_1_1) {
        /* Advertise available upgrade protocols */
        if (tls_enabled() &&
            config_mupdate_server && config_getstring(IMAPOPT_PROXYSERVERS)) {
            txn->flags.upgrade |= UPGRADE_TLS;
        }
        if (http2_enabled()) txn->flags.upgrade |= UPGRADE_HTTP2;

        if (txn->flags.upgrade) txn->flags.conn |= CONN_UPGRADE;
    }
```
vs. stable:
```c
    else if (!httpd_tls_done && txn->flags.ver == VER_1_1) {
        /* Advertise available upgrade protocols */
        txn->flags.conn |= CONN_UPGRADE;
        txn->flags.upgrade = UPGRADE_HTTP2;
        if (config_mupdate_server && config_getstring(IMAPOPT_PROXYSERVERS))
            txn->flags.upgrade |= UPGRADE_TLS;
```

As you can see, UPGRADE_HTTP2 is included, even when nghttp2 is not found.  Thus cyrus probably doesn't  send empty Upgrade: without the change above, but it is not supposed also to advertise HTTP/2 support.

I think using \#if HAVE_NGHTTP2 ... \#endif is better than `if http2_enabled()` because it produces smaller binaries and doesn't include code, that is supposed will never to be executed.

**Comment by elliefm** [comment]
#2691 has been merged to 3.0, but master is different enough that it's not a simple cherry-pick to get it there too

**Comment by dilyanpalauzov** [comment]
Just in case somebody is curious, I am not going to work on this on the master branch:

As a rule of thumb, I use version 3.0, which is understandable, and thus I provide corrections, that work towards that towards version.  There is nothing I can do for that fact that the discrepancies between the NEXT and CURRENT stable versions are that big, neither I see a need for a release cycle that leads to such big differences.

I worked on this in 2018.  I understood the specifics on this change good back then.  Now is 2019 and I have forgotten everything related.  So I have to get insight once again, completely from the beginning.  If the change was handled for the 3.0 branch in reasonable time after being submitted, I would not have forgotten all the details and it would have been much less efforts to get this on NEXT stable.
